core/blockly.js

```diff
@@ -273,6 +281,7 @@ Blockly.copy_ = function(toCopy) {
   }
   Blockly.clipboardXml_ = xml;
   Blockly.clipboardSource_ = toCopy.workspace;
+  Blockly.copiedObject_ = toCopy;
```

Comment:
Thought about it some more. Instead of putting the block stack to copy here can you put the map of typecounts? That way it'll be quick to check when pasting and you won't need to undo anything.

Reply:
Sounds like a good solution, but could you give me some tips on implimentation?

Currently the map of typecounts is stored as a local variable inside the isDuplicatable function. It could be moved to a property of the block object, but to get an updated version of it you'd need to call some getCopyableBlocksTypeCounts() function (since isDuplicatable stops once it runs into a type that goes over the remainingCapacityOfType).

So if the above doesn't have any gaps in logic, we have two options:
1) getCopyableBlocksTypeCounts() & isDuplicatable() share duplicate code.
2) isDuplicatable() calls getCopyableBlocksTypeCounts() and then iterates over the types (like blockly core would do on paste).

Idk if I should minimize iterations, duplicate code, or if there is a different solution.

Reply:
Let's go with 2. It'll be easier to maintain and visiting every block once shouldn't be too bad.

The method probably belongs in Blockly.utils.getBlockTypeCounts(block).